cmd/immuadmin/command/hot_backup.go

```diff
@@ -0,0 +1,661 @@
+var ErrTxNotInFile = errors.New("last known transaction not in file")
+
+type commandlineHotBck struct {
+	commandline
+}
+
+func newCommandlineHotBck(os immuos.OS) (*commandlineHotBck, error) {
+	cl := commandline{}
+	cl.config.Name = "immuadmin"
+	cl.context = context.Background()
+	cl.os = os
+
+	return &commandlineHotBck{cl}, nil
+}
+
+func (clb *commandlineHotBck) Register(rootCmd *cobra.Command) *cobra.Command {
+	clb.hotBackup(rootCmd)
+	clb.hotRestore(rootCmd)
+	return rootCmd
+}
+
+type backupParams struct {
+	output   string
+	startTx  uint64
+	append   bool
+	progress bool
+}
+
+func (cl *commandlineHotBck) hotBackup(cmd *cobra.Command) {
+	ccmd := &cobra.Command{
+		Use:   "hot-backup <db_name>",
+		Short: "Make a copy of the database without stopping",
+		Long: "Backup a database to file/stream without stopping the database engine. " +
+			"Backup can run from the beginning or starting from arbitrary transaction.",
+		PersistentPreRunE: cl.ConfigChain(cl.connect),
+		PersistentPostRun: cl.disconnect,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			file := io.Writer(os.Stdout)
+			params, err := prepareBackupParams(cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			udr, err := cl.immuClient.UseDatabase(cl.context, &schema.Database{DatabaseName: args[0]})
+			if err != nil {
+				return err
+			}
+			cl.context = metadata.NewOutgoingContext(cl.context, metadata.Pairs("authorization", udr.GetToken()))
+
+			if params.output != "-" {
+				f, err := cl.verifyBackupFile(params)
+				if err != nil {
+					return err
+				}
+				defer f.Close()
+				file = f
+			}
+
+			return cl.runHotBackup(file, params.startTx, params.progress)
+		},
+		Args: cobra.ExactArgs(1),
+	}
+	ccmd.Flags().StringP("output", "o", "-", "output file, \"-\" for stdout")
+	ccmd.Flags().Uint64("start-tx", 1, "Transaction ID to start from")
+	ccmd.Flags().Bool("progress", false, "show progress indicator")
+	ccmd.Flags().BoolP("append", "i", false, "append to file, if it already exists (for file output only)")
+	cmd.AddCommand(ccmd)
+}
+
+func prepareBackupParams(flags *pflag.FlagSet) (*backupParams, error) {
+	var params backupParams
+	var err error
+
+	params.output, err = flags.GetString("output")
+	if err != nil {
+		return nil, err
+	}
+	params.append, err = flags.GetBool("append")
+	if err != nil {
+		return nil, err
+	}
+	params.startTx, err = flags.GetUint64("start-tx")
+	if err != nil {
+		return nil, err
+	}
+	params.progress, err = flags.GetBool("progress")
+	if err != nil {
+		return nil, err
+	}
+
+	if params.startTx > 1 && params.append {
+		return nil, errors.New("don't use --append and --start-tx options together")
+	}
+
+	if params.output == "-" && params.append {
+		return nil, errors.New("--append option can be used only when outputting to the file")
+	}
+
+	return &params, nil
+}
+
+func (cl *commandlineHotBck) verifyBackupFile(params *backupParams) (*os.File, error) {
```

Comment:
shall verification be split ? as it's creating an empty file as secondary effect if the file does not exist?

Reply:
this function was renamed to make its purpose clear